### PR TITLE
Citieslist

### PR DESCRIPTION
**Pull request by fionamei**
Screen Shot 2022-07-05 at 1 36 22 PM
Screen Shot 2022-07-05 at 1 36 36 PM

### CODE REVIEW

File: ComfortZoneApp/app/src/main/java/com/example/comfortzone/WeatherDao.java [hunk 1]
```diff
@@ -14,10 +14,12 @@ public interface WeatherDao {
...
-//    List<WeatherDao> findByTemp(int temp);
-
     @Insert
     void insertAll(WeatherData... weatherData);
+
+    @Query("DELETE FROM WeatherData")
+    void deleteEntireTable();
+
+    @Query("SELECT * FROM WeatherData WHERE timeUploaded < :hourAgo")
+    List<WeatherData> getUploadTimes(long hourAgo);
```
Comment (kimjaspermui) [inline ComfortZoneApp/app/src/main/java/com/example/comfortzone/WeatherDao.java:25]: I think the method name is different than what it returns..I wonder if there's different SQL query such that you can simply check and return a boolean if there's one entry that has time uploaded within hr then you can probably just call getAll after, it might be more efficient that way? then you can name this method as `isUploadedWithinHrAgo` or `isWeatherDataFresh` and return boolean

File: city.json [hunk 2]
```diff
@@ -0,0 +1,1982 @@
+[
```
Comment (kimjaspermui) [inline city.json]: wait, why is this added back?

File: ComfortZoneApp/app/src/main/java/com/example/comfortzone/Fragments/FlightFragment.java [hunk 3]
```diff
@@ -23,6 +23,7 @@
 import com.google.gson.Gson;
 
 import java.util.List;
+import java.util.concurrent.TimeUnit;
 
 public class FlightFragment extends Fragment {
```
Comment (kimjaspermui) [inline ComfortZoneApp/app/src/main/java/com/example/comfortzone/Fragments/FlightFragment.java:29]: remember fragment/activities should only think about view as much as possible, think about how we can group related methods into different helper classes to help delegate the business logic

File: ComfortZoneApp/app/src/main/java/com/example/comfortzone/WeatherDao.java [hunk 4]
```diff
@@ -14,10 +14,12 @@ public interface WeatherDao {
...
-//    @Query("SELECT * FROM WeatherData WHERE `temp` > :temp")
-//    List<WeatherDao> findByTemp(int temp);
-
     @Insert
     void insertAll(WeatherData... weatherData);
+
+    @Query("DELETE FROM WeatherData")
+    void deleteEntireTable();
+
+    @Query("SELECT * FROM WeatherData WHERE timeUploaded < :hourAgo")
```
Comment (kimjaspermui) [inline ComfortZoneApp/app/src/main/java/com/example/comfortzone/WeatherDao.java:24]: I'm confused, according to your logic above, shouldn't this be `timeUploaded > :hourAgo` since hour ago should be less than timeUploaded for weather data to qualify?

File: ComfortZoneApp/app/src/main/java/com/example/comfortzone/Fragments/FlightFragment.java [hunk 5]
```diff
@@ -53,8 +54,8 @@ public void maybeUpdateCitiesList() {
         AllWeathersDatabase db = AllWeathersDatabase.getDbInstance(getContext().getApplicationContext());
         Long timeNow = System.currentTimeMillis();
         Long hourAgo = timeNow - TimeUnit.HOURS.toMillis(1L);
-        List<WeatherData> timesGreater = db.weatherDao().getUploadTimes(hourAgo);
-        if (timesGreater.isEmpty()) {
+        List<WeatherData> timesBeforeHour = db.weatherDao().getUploadTimes(hourAgo);
+        if (!timesBeforeHour.isEmpty()) {
```
Comment (kimjaspermui) [inline ComfortZoneApp/app/src/main/java/com/example/comfortzone/Fragments/FlightFragment.java:88]: i guess this also works (referring to my previous comment) - we are clearing only if we have local data before hour ago

File: ComfortZoneApp/app/src/main/java/com/example/comfortzone/Fragments/FlightFragment.java [hunk 6]
```diff
@@ -47,21 +51,49 @@ public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceStat
...
+        });
+    }
+
     public void maybeUpdateCitiesList() {
         AllWeathersDatabase db = AllWeathersDatabase.getDbInstance(getContext().getApplicationContext());
         Long timeNow = System.currentTimeMillis();
         Long hourAgo = timeNow - TimeUnit.HOURS.toMillis(1L);
         List<WeatherData> timesBeforeHour = db.weatherDao().getUploadTimes(hourAgo);
-        if (!timesBeforeHour.isEmpty()) {
+        if (db.weatherDao().getAll().isEmpty() || timesBeforeHour.isEmpty()) {
```
Comment (kimjaspermui) [inline ComfortZoneApp/app/src/main/java/com/example/comfortzone/Fragments/FlightFragment.java]: why do you need to check if the database is empty? also, if it's empty, do we still need to delete the table?

File: ComfortZoneApp/app/src/main/java/com/example/comfortzone/Fragments/FlightFragment.java [hunk 7]
```diff
@@ -47,21 +51,49 @@ public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceStat
...
             db.weatherDao().deleteEntireTable();
-            getCities();
+            updateCities();
         }
     }
 
-    public void getCities() {
+    public void updateCities() {
```
Comment (kimjaspermui) [inline ComfortZoneApp/app/src/main/java/com/example/comfortzone/Fragments/FlightFragment.java:93]: I think we are updating weather data rather than cities, just optional, but I think it's clearer if you refer them as weather. Additionally, since you are doing network call to get new weather data for the cities and also saving it, maybe a better method name would be `getAndStoreWeatherData`?

File: ComfortZoneApp/app/src/main/java/com/example/comfortzone/Fragments/FlightFragment.java [hunk 8]
```diff
@@ -87,7 +87,7 @@ public void maybeUpdateCitiesList() {
         Long timeNow = System.currentTimeMillis();
         Long hourAgo = timeNow - TimeUnit.HOURS.toMillis(1L);
         List<WeatherData> timesBeforeHour = db.weatherDao().getUploadTimes(hourAgo);
-        if (db.weatherDao().getAll().isEmpty() || timesBeforeHour.isEmpty()) {
+        if (db.weatherDao().getAll().isEmpty() || !timesBeforeHour.isEmpty()) {
```
Comment (kimjaspermui) [inline ComfortZoneApp/app/src/main/java/com/example/comfortzone/Fragments/FlightFragment.java]: I like how this line alone evolved many times XD

File: ComfortZoneApp/app/src/main/java/com/example/comfortzone/WeatherClient.java [hunk 9]
```diff
@@ -86,16 +86,13 @@ public void onResponse(@NonNull Call call, @NonNull Response response) throws IO
...
-        getCityData(new CityListCallback() {
-            @Override
-            public void cityList(List<City> cityList) {
-                List<String> ids = cityList.stream().map(city -> String.valueOf(city.getId())).collect(Collectors.toList());
-                List<List<String>> batchesCities = Lists.partition(ids, MAX_API_CITIES);
-                List<String> batchCityUrls = batchesCities.stream().map(batch -> getBatchUrl(batch)).collect(Collectors.toList());
-                callback.weatherUrlGroupIds(batchCityUrls);
-            }
-        });
+    public void getGroupWeatherUrl(Context context, GroupUrlCallback callback) {
```
Comment (kimjaspermui) [inline ComfortZoneApp/app/src/main/java/com/example/comfortzone/WeatherClient.java:89]: oh nice, you cleaned it up eventually..

File: ComfortZoneApp/app/src/main/java/com/example/comfortzone/Fragments/FlightFragment.java [hunk 10]
```diff
@@ -53,8 +53,10 @@ public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceStat
         client = new WeatherClient();
         cityList = new ArrayList<>();
         flightsAdapter = new FlightsAdapter(getContext(), cityList);
+        db = AllWeathersDatabase.getDbInstance(getContext().getApplicationContext());
 
         initViews(view);
+//        db.weatherDao().deleteEntireTable();
```
Comment (kimjaspermui) [inline ComfortZoneApp/app/src/main/java/com/example/comfortzone/Fragments/FlightFragment.java:59]: forgot to remove?

File: ComfortZoneApp/app/src/main/java/com/example/comfortzone/Fragments/FlightFragment.java [hunk 11]
```diff
@@ -68,33 +70,28 @@ private void initViews(@NonNull View view) {
...
+                    db.weatherDao().insertAll(weatherData);
+                }
+            });
+        }
         Long timeNow = System.currentTimeMillis();
         Long hourAgo = timeNow - TimeUnit.HOURS.toMillis(1L);
         List<WeatherData> timesBeforeHour = db.weatherDao().getUploadTimes(hourAgo);
-        if (db.weatherDao().getAll().isEmpty() || !timesBeforeHour.isEmpty()) {
-            db.weatherDao().deleteEntireTable();
+        if (!timesBeforeHour.isEmpty()) {
```
Comment (kimjaspermui) [inline ComfortZoneApp/app/src/main/java/com/example/comfortzone/Fragments/FlightFragment.java:88]: this is it?! 😆

File: ComfortZoneApp/app/src/main/java/com/example/comfortzone/Fragments/FlightFragment.java [hunk 12]
```diff
@@ -114,7 +111,13 @@ public void weatherData(String data) {
     public void saveCities(WeatherGroupData weathers) {
         WeatherData[] weatherData = weathers.getWeathers();
         AllWeathersDatabase db = AllWeathersDatabase.getDbInstance(getContext().getApplicationContext());
-        db.weatherDao().insertAll(weatherData);
+        for (WeatherData weather : weatherData) {
+            int id = weather.getId();
+            WeatherData weatherFromDb = db.weatherDao().getWeatherById(id);
+            weatherFromDb.setTempData(weather.getTempData());
+            weatherFromDb.setTimeUploaded(weather.getTimeUploaded());
+            db.weatherDao().updateWeatherData(weatherFromDb);
```
Comment (kimjaspermui) [inline ComfortZoneApp/app/src/main/java/com/example/comfortzone/Fragments/FlightFragment.java:119]: qq: is it guarantee that the weather data exist in local database?

**Review by kimjaspermui — APPROVED** [review]
really like how you learned and applied new tools into this project, it helped achieve our goal effectively while you are expanding your knowledge at the same time, good job!

File: ComfortZoneApp/app/src/main/java/com/example/comfortzone/WeatherClient.java [hunk 13]
```diff
@@ -86,22 +86,23 @@ public void onResponse(@NonNull Call call, @NonNull Response response) throws IO
...
-                }
+                List<String> batchCityUrls = batchesCities.stream().map(batch -> getBatchUrl(batch)).collect(Collectors.toList());
+                callback.weatherUrlGroupIds(batchCityUrls);
             }
         });
     }
 
+    private String getBatchUrl(List<String> batchCity) {
+        String idString = String.join(",", batchCity);
+        return String.format(API_GROUP_BASE_URL + "&id=%s", idString);
```
[the remark this answers is not in the document]
Comment (fionamei) [reply ComfortZoneApp/app/src/main/java/com/example/comfortzone/WeatherClient.java]: I tried using addQueryParameter, but because the string of ids need to be separated by commas, I wasn't able to use addQueryParameter since they turn all commas into a string of characters

File: ComfortZoneApp/app/src/main/java/com/example/comfortzone/AllWeathersDatabase.java [hunk 14]
```diff
@@ -20,7 +20,7 @@ public abstract class AllWeathersDatabase extends RoomDatabase {
     public static AllWeathersDatabase getDbInstance(Context context) {
         if (context == null || INSTANCE == null) {
             INSTANCE = Room.databaseBuilder(context.getApplicationContext(), AllWeathersDatabase.class, DB_NAME)
-                    .allowMainThreadQueries().build();
+                    .allowMainThreadQueries().fallbackToDestructiveMigration().build();
```
[the remark this answers is not in the document]
Comment (fionamei) [reply ComfortZoneApp/app/src/main/java/com/example/comfortzone/AllWeathersDatabase.java:23]: when i change database versions, room uses migrations to upgrade the version, but if theyre not found, the app will crash, so adding fallbackToDestructiveMigration recreates the table so that it wont crash
